packages/component/src/component_container.ts

```diff
@@ -0,0 +1,82 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Provider } from './provider';
+import { Component } from './component';
+import { Name } from './types';
+
+/**
+ * ComponentContainer that provides Providers for service name T, e.g. `auth`, `auth-internal`
+ */
+export class ComponentContainer {
+  private readonly providers = new Map<string, Provider>();
```

Comment:
When the types of values are different and known, instead of a `Map` I prefer using an object, which allows you to type the values independently. Something like:

```typescript
type ProviderMap = { [T in Name]?: Provider<T> };

const providers: ProviderMap = {};
```

And then you should be able to remove the casts in `getProvider`.

Reply:
Hmm, I got an error without casting - 
`'"test"' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '"test" | "badtest"'.`

`test` and `badtest` are the service names I defined and used for testing.

Reply:
Just tried it myself, I think the problem is that without casting TypeScript thinks the values of the object are `Provider<'test'> | Provider<'badTest'>` etc. instead of `Provider<'test' | 'badTest'>`, but I might be wrong.

Sorry about that, yeah I think we should keep the `Map` :sweat_smile: